SHARD-1386 : Harden relayer collector syncing

abdulazeem-tk4vr opened the pull request:
linear: https://linear.app/shm/issue/SHARD-1386/harden-relayer-collector-syncing
- Added data patcher to sync transactions lost between cycles
- Added code to avoid collector turning into zombie process upon errors being thrown - Stability
- Added checkpoint table to refer to the last known stable cycle of information in the collector dB

Glitch18 committed:
fix: improper error handling

Glitch18 committed:
feat: better server run structure

Glitch18 committed:
feat: create checkpoint database

abdulazeem-tk4vr committed:
feat : add checkpoint table fetch and insert function

abdulazeem-tk4vr committed:
feat : init services once and add patcher service to sync data

abdulazeem-tk4vr committed:
feat: implement checkpoint fetching and patching process in collector

Glitch18 committed:
feat: validate collector keypair configuration and ensure server starts asynchronously

abdulazeem-tk4vr committed:
feat : add retry mechanism for patching

abdulazeem-tk4vr committed:
feat : insert checkpoint after validating cycle data

abdulazeem-tk4vr committed:
feat : update table structure, busy_timeout and graceful shutdown

Glitch18 committed:
feat: add sigint handles

Glitch18 committed:
feat: full db rewrite to fix lock errors

aniketdivekar commented on package.json:
```diff
@@ -34,11 +32,14 @@
     "@types/socket.io": "2.1.11",
     "amqplib": "0.10.4",
     "axios": "1.4.0",
+    "better-sqlite3": "^11.8.1",
```
let's used fixed version

aniketdivekar commented on src/class/validateData.ts:
```diff
@@ -60,6 +61,11 @@ export async function validateData(data: Data): Promise<boolean> {
       data.cycle.counter + 1,
       data.cycle.cycleRecord.start + CONFIG.blockIndexing.cycleDurationInSeconds
     )
+    if (data.cycle.counter) {
```
as discussed, we need to remove this

Glitch18 committed:
feat: await cleanup

aniketdivekar commented on src/collector.ts:
```diff
@@ -376,11 +352,82 @@ const addSigListeners = (): void => {
 }
 
-startServer()
+const startLoop = async () => {
```
we should name it better (`startCollector` maybe)

aniketdivekar commented on src/collector.ts:
```diff
@@ -376,11 +352,82 @@ const addSigListeners = (): void => {
+      console.error(`Collector process stopped due to error: ${e.message}`)
+      console.log('Attempting fix..')
+      // fetch latest checkpoint
+      const lastKnownCycle = await checkpoint.fetchCheckpoint()
```
assuming that it's always a incorrect data issue is wrong I guess. we can check for specific error message/code maybe so we run patching only when its required

aniketdivekar commented on src/storage/checkpoint.ts:
```diff
@@ -0,0 +1,49 @@
+ */
+export async function insertCheckpoint(value: number, type: string = 'cycle'): Promise<void> {
+  try {
+    const sql = 'REPLACE INTO `checkpoint` (type, value) VALUES (?, ?)'
```
will REPLACE insert it if record doesn't exist?

abdulazeem-tk4vr replied:
we tried insert or replace and it was inserting new rows instead of replacing. Need to check this out

abdulazeem-tk4vr replied:
![image](https://github.com/user-attachments/assets/3bf400fb-09cd-4700-b863-fb1080556a3d)

aniketdivekar commented on src/storage/receipt.ts:
```diff
@@ -54,10 +55,27 @@ export async function bulkInsertReceipts(receipts: Receipt[]): Promise<void> {
     console.log(e)
+    const value = extractValuesFromArray(receipts.slice(0, 1))
+    const fields = Object.keys(receipts[0])
+    zip(fields, value)
```
what does this do? and also issue may not be in `receipts[0]` but in `receipts[5/10/N]`. 
(honestly, we can remove this code block)

abdulazeem-tk4vr replied:
@Glitch18 is better suited to resolve this

Glitch18 replied:
Done

aniketdivekar commented on src/storage/sqlite3storage.ts:
```diff
@@ -16,24 +15,26 @@ export interface DbOptions {
-      console.log('VERY SLOW QUERY', sql, time)
-    }
-  })
+  // NOT SUPPORTED IN BETTER-SQLITE3
```
what's an option for this? we may need to have some profiling

urnotsam replied:
we could make a db wrapper class than can proxy the run, prepare, all , etc calls to run the profiling.  that way we wouldnt' need to replace all usages of db.x instead could just import our wrapper class as db = new WrapperDatabase and everything should still work

abdulazeem-tk4vr replied:
maybe we address this in another ticket?

aniketdivekar commented on src/utils/patchCollector.ts:
```diff
@@ -0,0 +1,66 @@
+    try {
+      if (!endCycle) {
+        const response = await DataSync.queryFromDistributor(DataSync.DataType.TOTALDATA, {})
+        if (response.data && response.data.totalReceipts >= 0 && response.data.totalCycles >= 0) {
```
I think `response.data` has errors sometimes.. in else block or `if (endCycle === undefined) {` block, we should print response.data so that we know what exactly went wrong. 
(you can reproduce this by passing incorrect signature to distributor)

aniketdivekar commented on src/utils/patchCollector.ts:
```diff
@@ -0,0 +1,66 @@
+
+      await DataSync.downloadAndSyncGenesisAccounts() // To sync accounts data that are from genesis accounts/accounts data that the network start with
+      // TO DO : revisit purpose of genesis syncing
+      await DataSync.downloadCyclcesBetweenCycles(startCycle, endCycle, patchOnlyMissingData)
```
lets patch cycles at the end.(just making sure we have saved txs, receipts for a cycle before saving cycle itself)

Glitch18 committed:
feat: add cycle data handling and verification process

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+      // Start verification and checkpointing process here
+      // Check if we have cycle number 'checkPointer' in the db, if not, invoke patcher
+      checkPointer = await checkpoint.fetchCheckpoint();
+      const currentCycle = await cycle.queryCycleByCounter(checkPointer + 1)
```
```suggestion
      const nextCheckpointCycle = await cycle.queryCycleByCounter(checkPointer + 1)
```

urnotsam replied:
agree with suggestion but dont commit using github as the other references need to be updated as well

abdulazeem-tk4vr replied:
gotcha

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+      if (!currentCycle) {
+        throw Error(`Cycle ${currentCycle} is missing from the database.`)
+      }
+      while (endPointer < currentCycle.counter + 11) { // Wait till we have 11 cycles of data
```
please move `11` to a constant

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+      }
+      while (endPointer < currentCycle.counter + 11) { // Wait till we have 11 cycles of data
+        const latestCycle = await cycle.queryLatestCycleRecords(1);
+        if (latestCycle[0].counter >= endPointer + 10) {
```
```suggestion
        if (latestCycle[0].counter >= endPointer + 11) {
```

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+      await checkpoint.insertCheckpoint(checkPointer + 1)
+
+      // Verification failed
+      console.log(`🔴 Verification failed for cycle ${checkPointer + 1}. Checkpoint not updated.`)
```
probably not needed

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
-process.on('uncaughtException', (error) => {
-  console.error('Uncaught Exception in Collector: ', error)
-})
+startLoop()
```
`startCollectorServer`

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+    // Perform any necessary cleanup here
+    db.close() // Close the database connection
+    console.log('Collector shut down complete.')
+    process.exit(0) // Exit the process
```
```suggestion
    process.exit(1) // Exit the process
```

aniketdivekar commented on src/utils/patchCollector.ts:
```diff
@@ -0,0 +1,66 @@
+      // await DataSync.downloadAndSyncGenesisAccounts() // To sync accounts data that are from genesis accounts/accounts data that the network start with
+      // TO DO : revisit purpose of genesis syncing
+      await DataSync.downloadCyclcesBetweenCycles(startCycle, endCycle, patchOnlyMissingData)
+      if (config.verbose) console.log('Cycles Patched!')
```
we can remove verbose check

urnotsam commented on src/collector.ts:
```diff
@@ -381,11 +354,122 @@ const addSigListeners = (): void => {
+      }
+
+      // We should always have the next 11 cycles here. Fetch the data from distributor
+      const response = await DataSync.queryFromDistributor(DataSync.DataType.CYCLEDATA, { cycle: checkPointer + 1 })
```
usage of` checkPointer + 1` in main places below.  Make it a variable

abdulazeem-tk4vr committed:
fix: update better-sqlite3 version to use fixed version

abdulazeem-tk4vr committed:
fix : first iteration of review comments

abdulazeem-tk4vr committed:
fix :  second iteration of review comments

aniketdivekar commented on src/class/validateData.ts:
```diff
@@ -5,6 +5,7 @@ import { insertOrUpdateCycle } from '../storage/cycle'
 import { processReceiptData } from '../storage/receipt'
 import { processOriginalTxData } from '../storage/originalTxData'
 import { CycleLogWriter, ReceiptLogWriter, OriginalTxDataLogWriter } from './DataLogWriter'
+import * as checkpoint from '../storage/checkpoint'
```
this import isn't needed

Glitch18 committed:
fix: remove redundant error logging for bulkInsertReceipts

aniketdivekar commented on src/collector.ts:
```diff
@@ -381,11 +354,148 @@ const addSigListeners = (): void => {
+    console.log('Attempting fix..')
+
+    // Starts the syncing process from the beginning
+    const status = await startPatching(0)
```
this is not correct right? it should start syncing from whatever checkpoint we have in DB.
I also hope we considered syncing last 100 cycles by default on restart @abdulazeem-tk4vr

Glitch18 committed:
fix: remove unused checkpoint import in validateData

Glitch18 committed:
fix: start HTTP server when collector mode is set to MQ

abdulazeem-tk4vr committed:
fix: enhance checkpointing and patching process for improved data robustness

fix : add error logs

fix: ensure start cycle does not go below zero in patching process

aniketdivekar committed:
added some debug logs with prefix `[SHARD-1386]`

Glitch18 committed:
refactor: improve checkpoint handling and enhance variable naming for clarity

Glitch18 committed:
feat: implement CycleDataCache for efficient cycle data retrieval and caching

Glitch18 committed:
fix: add error handling for distributor queries and improve logging for missing cycles

Glitch18 committed:
fix: improve error logging in DataSync and patchCollector for better debugging

Glitch18 committed:
refactor: streamline RabbitMQ consumer management and enhance shutdown process

Glitch18 committed:
fix: add missing await for sleep

Glitch18 committed:
fix: broken distributor verifcation mechanism

aniketdivekar committed:
use txId & timestamp to create key for originalTxsMap. this will allow all data to be inserted in collector.

aniketdivekar committed:
fix: use originalTxsCount table count instead of transactions table count

abdulazeem-tk4vr committed:
fix : update rolling checkpoint window to be consistent with archiver

aniketdivekar committed:
chore: moved few logs behind verbose flag + added some informative logs

aniketdivekar committed:
improvement: immediate data verification for next checkpoint

devendra-shardeum commented on src/class/DataSync.ts:
```diff
@@ -77,8 +82,8 @@ export const queryFromDistributor = async (
-    console.log(`Error while querying ${url} for data ${data}`, e)
-    return null
+    console.log(`Error while querying ${url}: ${e.response.data.error}`)
+    return e.response
```
let's not return the error in response.. logging is enough imo

devendra-shardeum commented on src/class/DataSync.ts:
```diff
@@ -597,15 +614,31 @@ export const downloadReceiptsBetweenCycles = async (
   let endCycle = startCycle + 100
   for (; startCycle <= totalCyclesToSync; ) {
```
this can go in infinite loop.. better to have max retries logic here as well imo

devendra-shardeum commented on src/collector.ts:
```diff
@@ -383,43 +358,196 @@ const connectToDistributor = (): void => {
 
+const stopRMQEventsConsumers = async (): Promise<void> => {
+  console.log(`Initiated RabbitMQ connections cleanup`)
+  await rmqCyclesConsumer.cleanUp()
```
better to Run all cleanups in parallel using Promise.allSettled() imo
